### PR TITLE
fix: extract css in same order as imports

### PR DESCRIPTION
**Pull request by wardpeet**
### Description
When upgrading rollup to v2 and using [@rollup/plugin-node-resolve](https://www.npmjs.com/package/@rollup/plugin-node-resolve). My stylesheet wasn't deterministic and tests failed randomly. ModuleIds kept by Rollup are stored by resolve order, not import order. So, it depends on the time it takes to resolve a file.

I added a test where I resolving is done late but the css file still has the correct order.

Fixes https://github.com/egoist/rollup-plugin-postcss/issues/96

I've also made an issue in the rollup repository:
https://github.com/rollup/rollup/issues/3682

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -149,10 +172,15 @@ export default (options = {}) => {
 
         const concat = new Concat(true, fileName, '\n')
         const entries = [...extracted.values()]
-        const { modules } = bundle[normalizePath(path.relative(dir, file))]
+        const { modules, facadeModuleId } = bundle[
```
Comment (BPScott) [inline src/index.js:175]: `modules` here should already be in the deterministic import order, but in rollup v2 will exclude tree-shaken modules. 

If you're using rollup 2.21.0, we should be able to modify the return value of the transform hook, to ensure css code doesn't get shaken away and thus excluded from `modules`:

```diff
return {
        code: result.code,
        map: result.map || { mappings: '' },
        moduleSideEffects: 'no-treeshake',
}
```

I thiiiink that means we won't need to do this the sorting dance as the ordered list of css modules is something like the following for multiple bundles:

```
const moduleIds = flatMap(Object.values(bundle), (fileInfo) =>
  Object.keys(fileInfo.modules),
).filter(loaders.isSupported(id));
```

Comment (wardpeet) [reply src/index.js:175]: That would be amazing! I'll try to see if I can get it to work.

Comment (vladshcherbin) [reply src/index.js:175]: Seems like `modules` also don't work, they show files in random order for me in local rollup v2 tests

Comment (BPScott) [reply src/index.js:175]: @vladshcherbin could you define "unexpected order"? Could you provide an example of what your expected order would be?

Comment (vladshcherbin) [reply src/index.js:175]: @BPScott Sure, I've added a comment in https://github.com/rollup/rollup/issues/3682#issuecomment-667307596 with the order I'd expect to have. It's the same as @wardpeet mentioned as expected in linked issue.

Comment (wardpeet) [reply src/index.js:195]: I've enabled both methods now and added the no-treeshake for newer versions of rollup.

**Comment by wardpeet** [comment]
@egoist I based this PR on https://github.com/egoist/rollup-plugin-postcss/pull/299. If that one is okay, we could merge this one in after the rebase. If not i'll remove the prettier commit.

**Comment by BPScott** [comment]
Heck yeah I love that the bundle modules stuff seems to work! Fantastic work @wardpeet!

Having two ways of doing this does feel like a maintenance burden though, and might make other features like [multiple entry support](https://github.com/egoist/rollup-plugin-postcss/pull/289) tougher to implement.

@egoist how would you feel about if this only implemented using the bundle.modules/ no-treeshake approach and the claim this is a breaking change (and thus being a v4.0.0) as it requires >= rollup v2.21.0?

Then I think the multiple entry support would be pretty easy - change this from doing `flatMap` stuff to jam all files from all entries into a single list to something like (written but not ran).

```js
Object.entries(bundle).forEach(({bundleName, bundle}) => {
  // A list of all modules in this entrypoint  
  const moduleIds = bundle.modules.filter((id) => loaders.isSupported(id));
 
  // Do the stuff that emits out a css file for a given entrypoint
})
```

Which feels a lot easier than having to do all the existing logic for multi-entrys that have been attempted in https://github.com/egoist/rollup-plugin-postcss/pull/289 and prior multi-entry PRs

File: src/index.js [hunk 2]
```diff
@@ -15,15 +16,54 @@ function inferOption(option, defaultValue) {
...
+
+  return result
+}
+
+function flatMap(array, fn) {
+  if (Array.prototype.flatMap) {
+    return array.flatMap(fn)
+  }
+
+  let result = []
```
Comment (BPScott) [inline src/index.js:47]: When running on node 10 this causes the tests to fail for me. The following seems to work though:

```
function flatMap(array, fn) {
  if (Array.prototype.flatMap) {
    return array.flatMap(fn)
  }

  return array.reduce((memo, item) => memo.concat(fn(item)), []);
}
```

**Comment by BPScott** [comment]
Heya @wardpeet, me again. I spotted some awkwardness in the snapshots and as I had an idea what was up I figured I'd make a PR to your PR to demo the fix rather than try and explain it in comments: https://github.com/wardpeet/rollup-plugin-postcss/pull/1

File: src/index.js (hunk 2 above)
Comment (wardpeet) [reply src/index.js:47]: Thanks for spotting, I fixed it. I rather not use reduce as it's harder to reason with + `fn.call` mimics the same callback as flatMap would, that's why I used it.

**Comment by wardpeet** [comment]
@BPScott I do agree that the changes you made remove lots of the $undefined variables that are generated in the newer rollup version. The reason why I didn't look more into it, would mean this would become a breaking change and harder to get reviewed/merged. And terser will remove any of these undefined variables anyway so there is no perf impact except a more verbose output when not using terser.

If I move back to my first implementation this would be solved as well. We could land your changes after this one and see if we can bump it as a major but for now I just want myself unblocked 😄

**Comment by BPScott** [comment]
> If I move back to my first implementation, this would be solved as well. We could land your changes after this one and see if we can bump it as a major version. For now, I want myself unblocked. 😄

Yep, agree that I'm kinda derailing this PR and getting this behaviour working is top priority :)  I'm very happy to say this is good to ship and I'll poke at improvements with potential for breaking changes in a follow-up PR :shipit:

**Comment by wardpeet** [comment]
I've run tests on my mac but I couldn't reproduce the failure. Unsure what's happening here. I'll cleanup the PR and hopefully, this can get merged soon after.

**Comment by wardpeet** [comment]
I converted everything to the first version without any rollup changes needed. This should be good to go.

**Comment by wardpeet** [comment]
@egoist any chance we could merge this in? :) Happy to adjust the PR with feedback

**Review by egoist — APPROVED** [review]

**Pull request merged**